src/Education/XamlGenerator/IRuleHelpXamlBuilder.cs

```diff
@@ -0,0 +1,472 @@
+
+                    PushOutputElementInfo("ul", false);
+
+                    break;
+
+                case "table":
+                    WriteBlockElementStart("Table");
+                    writer.WriteAttributeString("BorderThickness", "1,1,1,1");
+                    writer.WriteAttributeString("BorderBrush", "Black");
+
+                    PushOutputElementInfo("table", false);
+
+                    break;
+
+                case "colgroup":
+                    writer.WriteStartElement("Table.Columns");
+                    PushOutputElementInfo("colgroup", false);
+
+                    break;
+
+                case "col":
+                    // This is an empty element, so there is nothing to push onto the stack.
+                    WriteEmptyElement("TableColumn");
+
+                    break;
+
+                case "thead":
+                    writer.WriteStartElement("TableRowGroup");
+                    writer.WriteAttributeString("Background", "LightGray");
+                    writer.WriteAttributeString("FontWeight", "Bold");
+
+                    PushOutputElementInfo("thead", false);
+
+                    break;
+
+                case "tr":
+                    writer.WriteStartElement("TableRow");
+                    tableAlternateRow = !tableAlternateRow;
+
+                    PushOutputElementInfo("tr", false);
+
+                    break;
+
+                case "th":
+                    writer.WriteStartElement("TableCell");
+                    writer.WriteAttributeString("BorderThickness", "0,0,0,1");
+                    writer.WriteAttributeString("BorderBrush", "Black");
+                    PushOutputElementInfo("th", false);
+
+                    break;
+
+                case "tbody":
+                    tableAlternateRow = true;
+                    writer.WriteStartElement("TableRowGroup");
+                    PushOutputElementInfo("tbody", false);
+
+                    break;
+
+                case "td":
+                    writer.WriteStartElement("TableCell");
+                    PushOutputElementInfo("td", false);
+
+                    if (tableAlternateRow)
+                    {
+                        writer.WriteAttributeString("Background", "LightBlue");
+                    }
+
+                    break;
+
+                default:
+                    Debug.Fail("Unexpected element type: " + reader.Name);
+                    writer.WriteStartElement(reader.Name);
+                    writer.WriteEndElement();
+                    break;
+            }
+        }
+
+        private void WriteEmptyElement(string name, string value = null)
+        {
+            Debug.Assert(reader.IsEmptyElement);
+            writer.WriteElementString(name, value);
+        }
+
+        private void WriteInlineElementStart(string elementName)
+        {
+            // If we are writing an inline element, we need a parent element that supports inlines.
```

Comment:
Long comments in this and the following methods to explain the one complicate part of building the XAML tree.

In summary: most XAML document elements can be classified as either `Inline` or `Block` (there are a few exceptions e.g. the table and list classes).

Some elements can't have child elements e.g. `LineBreak`.
Some elements can have one or more `Inline` children e.g. `Span` and `Paragraph`
Some elements can have one or more `Block` children e.g. `Section`, `ListItem`, `TableCell`.


Before adding a child element, we need to make sure it's parent can accept a child element of that type
e.g.
* current = `Paragraph`, want to add text -> no problem. `Paragraph` has `Inline` children, text is an `Inline`

* current = `ListItem`, want to add text -> can't. `ListItem` has `Blocks` as children. So, we first add a `Paragraph` (a block) -> `ListItem` -> `Paragraph` -> text.

* current = `Paragraph`, want to add a `List` -> can't. `List` is a `Block`. So, we need to walk back up the list of added elements closing them until we find something that accepts a `Block`. Note: there are limits on how far we can walk back up - we can only walk as far as an element that corresponds directly to an `HTML` opening tag. If we walk any further, we'll have a problem when we try to process the `HTML` element closing tag. To put it another way, we can only close extra intermediate elements we have added that don't map directly to an `HTML` tag.